### ISSUE TITLE
Allow metadata to be added to Task.async and friends

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/elixir#11716]
If you return a `Task` to be awaited on, then it is impossible to identify later on what created that task. We should support a `:metadata` field in the `Task` struct with an equivalent `Task.async(fun, metadata: ...)` API on all functions that return `%Task{}`.

**Comment by whatyouhide** [comment]
@josevalim when you say

> it is impossible to identify later on

can't you always pass around some metadata alongside the task to whoever you're giving the task to await on to? 🤔

**Comment by josevalim** [comment]
@whatyouhide the issue is that some libraries expect you to return a list of tasks that it will await on. We could change those libraries to allow list of tuples, where the task is the first element, but it is not very extensible as we would need to change all libraries where this occurs.

Another option would be for us to change Task.await and friends to include PID metadata in the error message (current stacktrace perhaps?). That could help us identify the tasks and without changing the API. Perhaps we should start here?

**Comment by whatyouhide** [comment]
@josevalim yeah that seems like a more incremental approach so that we don't have to change the API yet when the use case is still "fuzzy". I’m on board with that 🙃
